libsharedringbuffer/include/shared_ringbuffer/shared_ringbuffer.h

```diff
@@ -0,0 +1,241 @@
+ *                    0 inidicates they do not. Only one side of the shared memory regions needs to do this. 
+ */
+void ring_init(ring_handle_t *ring, ring_buffer_t *avail, ring_buffer_t *used, notify_fn notify, int buffer_init);
+
+/**
+ * Check if the ring buffer is empty.
+ *
+ * @param ring ring buffer to check.
+ *
+ * @return true indicates the buffer is empty, false otherwise.
+ */
+static inline bool ring_empty(ring_buffer_t *ring) 
+{
+    //return !((ring->write_idx - ring->read_idx) % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT);
+    return (ring->write_idx == ring->read_idx);
+}
+
+/**
+ * Check if the ring buffer is full
+ *
+ * @param ring ring buffer to check.
+ *
+ * @return true indicates the buffer is full, false otherwise.
+ */
+static inline bool ring_full(ring_buffer_t *ring)
+{
+    //return !((ring->write_idx - ring->read_idx + 1) % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT);
+    return CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT == (ring->write_idx - ring->read_idx);
+}
+
+/**
+ * Notify the other user of changes to the shared ring buffers.
+ * 
+ * @param ring the ring handle used.
+ *
+ */
+static inline void notify(ring_handle_t *ring) 
+{
+    if (ring->notify != NULL) {
+        return ring->notify();
+    }
+}
+
+/**
+ * Enqueue an element to a ring buffer
+ *
+ * @param ring Ring buffer to enqueue into.
+ * @param buffer address into shared memory where data is stored.
+ * @param len length of data inside the buffer above.
+ * @param cookie optional pointer to data required on dequeueing. 
+ *
+ * @return -1 when ring is empty, 0 on success.
+ */
+static inline int enqueue(ring_buffer_t *ring, uintptr_t buffer, size_t len, void *cookie) 
+{
+    if (ring_full(ring)) {
+        ZF_LOGE("Ring full");
+        return -1; 
+    }
+
+    buff_desc_t *buff_desc = &(ring->buffers[ring->write_idx % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT]);
+    buff_desc->encoded_addr = buffer;
+    buff_desc->len = len;
+    buff_desc->cookie = cookie;
+    ring->write_idx++;
+
+    /* Ensure data is written back to memory */
+    THREAD_MEMORY_RELEASE();
+
+    return 0;
+}
+
+/**
+ * Dequeue an element to a ring buffer.
+ *
+ * @param ring Ring buffer to Dequeue from.
+ * @param buffer pointer to the address of where to store buffer address.
+ * @param len pointer to variable to store length of data dequeueing.
+ * @param cookie pointer optional pointer to data required on dequeueing. 
+ *
+ * @return -1 when ring is empty, 0 on success.
+ */
+static inline int dequeue(ring_buffer_t *ring, uintptr_t *addr, size_t *len, void **cookie) 
+{
+    if (ring_empty(ring)) {
+        ZF_LOGF("Ring is empty");
+        return -1;
+    }
+
+    buff_desc_t *buff_desc = &(ring->buffers[ring->read_idx % CONFIG_LIB_SHARED_RINGBUFFER_DESC_COUNT]);
+    *addr = buff_desc->encoded_addr;
+    *len = buff_desc->len;
+    *cookie = buff_desc->cookie;
+
+    /* Ensure data is written back to memory */
```

Comment:
nitpicking: I think all we ensure here is, that no writes are reorderd beyond this point, so they are visible. But it's not guaranteed the data really ends up in memory, as this is no cache flush.